syntax = "proto3";

package sentry_protos.snuba.v1;

import "google/protobuf/timestamp.proto";
import "sentry_protos/snuba/v1/request_common.proto";
import "sentry_protos/snuba/v1/trace_item_attribute.proto";
import "sentry_protos/snuba/v1/trace_item_filter.proto";

enum Function {
  FUNCTION_UNSPECIFIED = 0;
  FUNCTION_SUM = 1;
  FUNCTION_AVERAGE = 2;
  FUNCTION_COUNT = 3;
  FUNCTION_P50 = 4;
  FUNCTION_P90 = 5;
  FUNCTION_P95 = 6;
  FUNCTION_P99 = 7;
  FUNCTION_AVG = 8;
  FUNCTION_MAX = 9;
  FUNCTION_MIN = 10;
  FUNCTION_UNIQ = 11;
}

message TimeSeriesRequest {
  RequestMeta meta = 1;
  TraceItemFilter filter = 2;
  repeated AttributeAggregation aggregations = 3;
  uint64 granularity_secs = 4;
  repeated AttributeKey group_by = 5;
}

message AttributeAggregation {
  Function aggregate = 1;
  AttributeKey key = 2;
  string label = 3;
}

message TimeSeries {
  string label = 1;
  // the names and values of the attribute keys which
  // were in the group by
  map<string, string> group_by_attributes = 2;
  // time bucket for each data point in the timeseries
  repeated google.protobuf.Timestamp buckets = 3;
  repeated float data_points = 4;
  // number of events used to calculate this timeseries
  uint64 num_events = 5;
  float avg_sampling_rate = 6;
}

message TimeSeriesResponse {
  repeated TimeSeries result_timeseries = 1;
}
